### ISSUE TITLE
[Suggestion] On demand source buffer?

### ISSUE THREAD

**Issue by shangjiaxuan** [kcat/openal-soft#325]
Is there any plans to implement a software synthesizer when OpenAL needs a source? 

Similar to the current use case of using a loopback device to implement custom backend, maybe there can be a extension that registers a function to be called when the data is needed? (Of course, also need some non-block guarantee here.) 

This can reduce the resource usage when playing compressed long audio without the need for extra memory copying and usage (e.g. plugging a decoder via a user pointer, and requesting a number of frames on the interface) compared to buffer queuing (and also more flexible if the user want a dynamic software synthesizer as source (also less latency when user wants to change the buffer contents)).

Maybe some interface like this to bind a generator to a buffer?
`void alBufferFuncSOFT(ALuint abo, ALenum format, ALsizei freq, alSourceFunc_t func, ALvoid* usr_ptr)`
`typedef Alsizei(*alSourceFunc_t)(Aluint abo, ALvoid* to_fill, ALuint min_frames, ALuint max_frames, ALvoid* usr_ptr /*, ALuint alignment, Alenum format, ...*/)`

(There may need to be format conversion set up after each call to callback, so maybe setting up format beforehand or providing more parameters to the callback and avoiding this conversion may be better but will require more effort on the user side).

The format, sample rate, etc can still be accessed the `alBufferxx`, `alGetBufferxx` interface, and buffer queuing and other things will still work, since it's still basically just a buffer on demand.

(If there is any plan on implementing this, I'm happy to dig down to the source and help...)

**Comment by shangjiaxuan** [comment]
Also, currently the `BackendBase::mMutex` is only called by the `BackendBase::lock()` virtual method, which in turn is only called in every buffer write callback (and a few functions to retrieve backend status). It is possible to remove the need to lock it entirely?

**Comment by kcat** [comment]
> This can reduce the resource usage when playing compressed long audio without the need for extra memory copying and usage (e.g. plugging a decoder via a user pointer, and requesting a number of frames on the interface)

This would actually be trickier than it may seem. Unless you know the exact behavior of a given decoder, you can't be sure that the decode function won't employ some locks (to ensure safe multi-threaded access to internal state) or allocations (for on-demand internal storage). With callbacks invoked by the mixer, you shouldn't be doing these things as they can stall processing without warning and make the audio update run late. Made worse by the fact that some backends may be more sensitive to this kind of thing than others, so you may not see a problem until a user with a particular system or configuration tries it.

There is, however, a currently-in-development extension for mapping a buffer's memory. This would allow avoiding extra copies where possible, since rather than using some temporary memory to write decoded frames to before calling `alBufferData` to copy it in, you can map the buffer and decode directly into it. It would still require a source buffer queue or something to stream it, though.

> Also, currently the BackendBase::mMutex is only called by the BackendBase::lock() virtual method, which in turn is only called in every buffer write callback (and a few functions to retrieve backend status). It is possible to remove the need to lock it entirely?(perhaps use a command buffer and let the caller wait on an update when the time frame is really small?)

I would like to get rid of that lock, but there's currently two major things that need that kind of synchronization: source state changes (`alSourcePlay[v]`, `alSourceStop[v]`, etc) and setting a playback offset. The exact behavior of these functions depends on the current state of a given source at the time the change is applied, while they're being asynchronously changed by the mixer (with the `alSource*v` functions also guaranteeing multiple sources change atomically, meaning multiple sources start/stop at exactly the same time). The caller must also always see state consistent with the given changes afterward.

This creates the challenge of synchronizing changes across multiple sources, depending on each source's current state at that same synchronization point (and ensuring the mixer doesn't change the state in the mean time). The state of all sources must also remain coherent, so if the app queries anything afterward it won't see temporarily incorrect or old state, it must always be "correct" for the change made.

There is likely a way to do this, I just haven't worked it out yet.

**Comment by shangjiaxuan** [comment]
Yes, I totally missed the lock guard at the play functions whens searching for calls to the lock method... Yes, that would be a problem if the the access is not synchronized. Are the other functions lock free within? If so, is making a fixed-length command queue (typically lock-free queues never blocks the worker thread.) where the caller thread blocks until the audio thread updates the state? (The when the queue is full, the caller thread just wait a little more to push the buffer and wait.)

This may potentially make the caller thread wait more time until an update from audio thread is issued (up to the time of update interval of the backend buffer, if that interval is more than 20ms, video playing on the same thread would certainly experience frame drops at 60fps.), but seems to be able to remove the lock there.

(Also, considering the use case of playing a source that dynamically changes, eg. a midi instrument from sofware synthesizer, or if the user would like to just plug in AL to their own mixing or synthesizing functions as output (similar to the use of alcRenderSamplesSOFT for plugging AL as input end), a callback can still be quite handy provided the user guarantees the function not to block.)

**Comment by kcat** [comment]
> Yes, I totally missed the lock guard at the play functions whens searching for calls to the lock method... Yes, that would be a problem if the the access is not synchronized. Are the other functions lock free within?

With respect to the backend/mixer lock, they should be. `alSourcePlay[v]`, `alSourcePause[v]`, `alSourceStop[v]`, `alSourceRewind[v]`, as well as two other places (within `SetSourcefv` and `SetSourceiv`, where they call `ApplyOffset`), are the only places I can think of that need that lock. Technically `FreeSource` does too, but that's just the same as `alSourceStop`, to ensure the source is stopped before destructing it.

> If so, is making a fixed-length command queue (typically lock-free queues never blocks the worker thread.) where the caller thread blocks until the audio thread updates the state?

Main problem is it could be a while for the mixer/worker thread to do its work. If your timing is unlucky, or a user's settings increase the update size, it can be over 20ms for another mix to occur. There may not even be another mix while the call is waiting (if the app is using a loopback device synchronously, mixing on the same thread as other AL calls, or if a normal device's playback is currently paused with `alcDevicePauseSOFT`).

At most, an AL call can wait for any asynchronous mix that's currently happening to finish. But it can't wait for another mix to happen, since there's no way to know when (or if) that will be.

> (Also, considering the use case of playing a source that dynamically changes, eg. a midi instrument from sofware synthesizer, or if the user would like to just plug in AL to their own mixing or synthesizing functions as input to AL (similar to the use of alcRenderSamplesSOFT for plugging AL to elsewhere), a callback can still be quite handy provided the user guarantees the function not to block.)

That guarantee is the tricky part, since there's no way for OpenAL Soft to verify a callback can never block (or otherwise take too long) until it does. I could only trust the developer doesn't do anything improper, even if it may otherwise work for them with their testing.

In that case though, that I do make the warning clear and concise and simply trust it's used right, there's one more issue of resampling. To resample, OpenAL Soft needs some previous samples (which it handles itself, holding on to what it had from last time) as well as some future samples. For static/non-streaming sources, it has all the samples it needs since the buffer's contents are completely fixed when playing. With streaming sources, it's assumed enough samples have been queued to avoid the risk of an underrun. So in both cases, it can read some samples beyond the playback offset.

With a callback, there would need to be a way to get enough samples to reach the new playback offset, plus extra samples beyond that point. The only way to avoid this would be to add an extra delay to callback sources, making playback some unknown number of samples behind the last samples given and not synchronize with non-callback sources (there will be some unknown number of samples difference between them even if they're started together with `alSourcePlayv`).

**Comment by shangjiaxuan** [comment]
I'm looking into a possible dirty implementation of the callback, but currently things seems a bit glitchy and many things are not taken care of (For example, setting the callback function pointer back to nullptr when reusing the buffer object, since being callback is currently implemented by checking by that pointer).

Also, there seems to be glitches currently with reach render's boundaries on callback buffers (clicking sound). Playing another looping sound at the same frequency does not induce same clicking on the other source.

Will look into it tomorrow... (Current proposed changes [here](https://github.com/shangjiaxuan/openal-soft/commit/669a7c633d26f552b0095253bcc1b1a750821467), naming style is a bit dirty for now...)

**Comment by kcat** [comment]
I'm not sure putting the callback on the buffer makes the most sense. That would cause a buffer to have an indefinite length, which creates confusion for things like looping and queueing. I think it would make more sense putting the calback directly onto the source instead, basically giving another source type: `AL_UNDETERMINED`, `AL_STATIC`, `AL_STREAMING`, and `AL_CALLBACK_SOFT`.

> Also, there seems to be glitches currently with reach render's boundaries on callback buffers (clicking sound).

It looks like you're not accounting for the future samples used for resampling. When the mixer prepares samples for processing, it first loads previous samples it stored from last time, then it loads samples  from the buffer up to the new source offset, then it loads some extra samples from the buffer that the resampler will want. Those extra samples will be loaded from the buffer again in the next mix.

However, with the callback only going forward in time, those extra samples it gets for one mix will be skipped over on the next mix, causing discontinuity and an audible click. The way to fix this would be to store those extra samples to reuse on the next mix, although this will cause a slight delay in callback-based sources that static or streaming sources won't have.

**Comment by shangjiaxuan** [comment]
The reasoning on making callback a kind of buffer is that it's basically a buffer on demand, and the user may still queue the buffer and use it as static. The indicator of termination, however, would be that the returned number of samples (the return value) is smaller than the requested value (the value passed into the callback)(currently it's size of filled, though). This could avoid the frequent use of `alSourcePlay` and `alSourceStop`, since it's still the same source.

But I think you are right, implementing this as a separate kind of source would be much more straightforward and much less work to do with existing code.

**Comment by shangjiaxuan** [comment]
I still don't quite understand how the re-sampling padding interact with data loaded from buffer...

From what I see, the `MixVoice` function reads from the stored last few samples from previous loop, and load the remaining source buffer with the needed data. Then the needed data is copied to the "last few samples" section. I don't see how the same section of a non-looping buffer needs to be loaded many times.

Is there something I was missing from the `DataPosInt` and `DataPosFrac` variables?

**Comment by kcat** [comment]
> I still don't quite understand how the re-sampling padding interact with data loaded from buffer...
>
> From what I see, the MixVoice function reads from the stored last few samples from previous loop, and load the remaining source buffer with the needed data. Then the needed data is copied to the "last few samples" section. I don't see how the same section of a non-looping buffer needs to be loaded many times.

The samples loaded for mixing are comprised of three parts. Past samples (those before the current source position), current samples (those between the current source position (mix start pos) and the destination source position (mix end pos)), and future samples (those after the destination source position). Resampling needs past and future samples to work correctly, or else a delay is needed to compensate for not having future samples.

The past samples are loaded from what it had saved last time. Then the current and future samples are loaded from the buffer. The future samples of one mix will be the current samples in the next mix, meaning those samples are read twice in succession. With normal sources this is fine since the buffer won't go anywhere while the source is using it. However, for a callback that expects to give samples only once this creates a problem, since if the callback gives future samples in one mix, it won't give them again as current samples in the next mix.

The way to fix it would be to also save the future samples and reuse them as current samples next time. This would effectively create the aforementioned delay, since the samples given by the callback will always be a little bit ahead of what's been mixed. The good news is, `MixVoice` already saves the future samples. It just doesn't use them as long as there's still a buffer to get real samples from, to avoid unnecessary delays. A callback-based path could easily load them as needed, as long as the delay is acceptable.

**Comment by shangjiaxuan** [comment]
Thank you SO much, the callback worked smoothly after making the callback case copy the whole re-sampling part (twice the padding). The new latency induced seemed reasonable enough (padding was 24 floating point mono samples, thus less than 1ms on 48kHz), the backend's time frame is still the major  source.

I also made a working demo of using the callback [here](https://github.com/shangjiaxuan/openal-soft/commit/2dce6c058a369d3e7cf98bd1da6ffbc55aa81dcf). The demo depends on `glfw` for keyboard input and some `opengl` calls (`SwapBuffers`) (The color in the window is dependent on the key pressed). Latency seems to be acceptable with default `wasapi` backend.

There is still much work to do to integrate this into existing code (e.g. checking this in incompatible calls, like making a buffer loop, or mapping buffers). I will look into that later.

I also made a change in `ALsource::~ALsource`'s loop of destroying the `ALbufferlistitem* queue`, since I was getting segfault by accessing a nullptr's members. I was wondering if this breaks anything.

Original was

    ALbufferlistitem *BufferList{queue};
    while(BufferList != nullptr)
    {
        std::unique_ptr<ALbufferlistitem> head{BufferList};
        /* If (BufferList->next==nullptr) */
        BufferList = head->mNext.load(std::memory_order_relaxed);
        /* the following will segfault */
        if(ALbuffer *buffer{BufferList->mBuffer}) DecrementRef(buffer->ref);
    }
    queue = nullptr;

Changed to:

    ALbufferlistitem *BufferList{queue};
    while(BufferList != nullptr)
    {
        std::unique_ptr<ALbufferlistitem> head{BufferList};
        if(ALbuffer *buffer{BufferList->mBuffer}) DecrementRef(buffer->ref);
        BufferList = head->mNext.load(std::memory_order_relaxed);
    }
    queue = nullptr;

I think I will be busy in the following few weeks, so I guess that will take some time before actually finished.

Thanks again for helping, couldn't have found out the reason myself.

(Attached is the sample compiled with VS2017 community IDE, in case you don't have glfw and glew installed.)

[CallbackTest.zip](https://github.com/kcat/openal-soft/files/3512849/CallbackTest.zip)
